### DISCUSSION TITLE
self.next_functions = [],  is declared but not not sure that its being used anywhere

### DISCUSSION THREAD

**Discussion by AmirAlasady** [harvard-edge/cs249r_book#1122] [TinyTorch Q&A] [answered]
in the Function class as shown: 
class Function:
    """
    Base class for differentiable operations.

    Every operation that needs gradients (add, multiply, matmul, etc.)
    will inherit from this class and implement the apply() method.

    **Key Concepts:**
    - **saved_tensors**: Store inputs needed for backward pass
    - **apply()**: Compute gradients using chain rule
    - **next_functions**: Track computation graph connections

    **Example Usage:**
    ```python
    class AddBackward(Function):
        def apply(self, grad_output):
            # Addition distributes gradients equally
            return grad_output, grad_output
    ```
    """

    def __init__(self, *tensors):
        """
        Initialize function with input tensors.

        Args:
            *tensors: Input tensors that will be saved for backward pass
        """
        self.saved_tensors = tensors
        self.next_functions = []

        # Build computation graph connections
        for t in tensors:
            if isinstance(t, Tensor) and t.requires_grad:
                # Check if this tensor was created by another operation
                # _grad_fn is only present if autograd is enabled and tensor came from an operation
                if getattr(t, '_grad_fn', None) is not None:
                    self.next_functions.append(t._grad_fn)

    def apply(self, grad_output):
        """
        Compute gradients for inputs.

        Args:
            grad_output: Gradient flowing backward from the output

        Returns:
            Tuple of gradients for each input tensor

        **Must be implemented by subclasses**
        """
        raise NotImplementedError("Each Function must implement apply() method")

we are declaring next_functions but i dont think its being used anywhere in the system later 
like if we inspect the backward method in the tensor class 
where we are supposed to trigger the graph traversal and execution 
as i was reading the book it hinted at the we will make use of this next_functions variable when we traverse in reversed topological order .. yet in the backward method 
we are not using this in any how as shown : 

    def backward(self, gradient=None):
        """
        Compute gradients via backpropagation.

        This is the key method that makes training possible!
        It implements reverse-mode automatic differentiation.

        **Algorithm:**
        1. Initialize gradient if not provided (for scalar outputs)
        2. Accumulate gradient in self.grad
        3. If this tensor has a _grad_fn, call it to propagate gradients
        4. Recursively call backward() on parent tensors

        **Example:**
        ```python
        x = Tensor([2.0], requires_grad=True)
        y = x * 3
        y.backward()  # Computes gradients for x
        print(x.grad)  # [3.0]
        ```
        """
        # Ensure gradient attributes exist
        _ensure_grad_attrs(self)

        # Only compute gradients if required
        if not _get_requires_grad(self):
            return

        # Initialize gradient if not provided (for scalar outputs)
        if gradient is None:
            if self.data.size == 1:
                gradient = np.ones_like(self.data)
            else:
                raise ValueError(
                    f"backward() called on non-scalar tensor without gradient argument.\n"
                    f"  Tensor shape: {self.shape}\n"
                    f"  Issue: For non-scalar outputs, you must provide the gradient from the next layer.\n"
                    f"  Fix: Call backward(gradient) with the gradient tensor from the loss function."
                )

        # Initialize or accumulate gradient
        if self.grad is None:
            self.grad = np.zeros_like(self.data)

        # Handle broadcasting: sum gradient to match self.data shape
        # This happens when operations broadcast tensors (e.g., adding bias to batch)
        if gradient.shape != self.grad.shape:
            # Step 1: Remove extra leading dimensions added during forward pass
            # Example: gradient (batch_size, features) → self.grad (features,)
            while gradient.ndim > self.grad.ndim:
                gradient = gradient.sum(axis=0)

            # Step 2: Sum over dimensions that were size-1 in original tensor
            # Example: bias with shape (1,) broadcast to (batch_size,) during forward
            for i in range(gradient.ndim):
                if self.grad.shape[i] == 1 and gradient.shape[i] != 1:
                    gradient = gradient.sum(axis=i, keepdims=True)

        self.grad += gradient

        # Propagate gradients through computation graph
        # _grad_fn is set by autograd enhancement when tensor is created from an operation
        grad_fn = getattr(self, '_grad_fn', None)
        if grad_fn is not None:
            grads = grad_fn.apply(gradient)

            # Recursively call backward on parent tensors
            for tensor, grad in zip(grad_fn.saved_tensors, grads):
                if isinstance(tensor, Tensor) and tensor.requires_grad and grad is not None:
                    tensor.backward(grad)

pls can someone explain to me what's the point of this then

**Comment by profvjreddi** [comment, marked as answer]
Great catch, @AmirAlasady. You are right. `next_functions` is declared but never used.

**Why it was originally included**

The `next_functions` pattern mirrors how PyTorch represents the computation graph internally. In PyTorch autograd, each node keeps explicit references to the gradient functions of its inputs. That structure enables things like graph traversal and topological sorting for non recursive backward passes.

When building TinyTorch, I initially included this while going back and forth on the design. At the time, I thought we might want explicit graph edges for a few reasons:

* Supporting non recursive backward passes via topological sort
* Exploring more complex gradient checkpointing ideas
* Staying closer to PyTorch’s internal structure for teaching purposes

**Why it should be removed**

For an educational implementation, the simpler recursive approach using `saved_tensors` is a better fit.

1. It is easier to understand. You can follow `backward()` calling `apply()` and then recursing backward without first learning a separate graph structure.
2. It reduces cognitive load. Every line of code should teach something, and unused code just creates confusion, as this question nicely surfaced.
3. It works correctly. The recursive approach already handles all of our current use cases.

I just opened PR #1123 to remove it.

Thanks for the careful reading of the code. This kind of attention to detail is exactly how you build strong engineering instincts.

**Reply by AmirAlasady** [reply]
thank you for that sir.

**Comment by profvjreddi** [comment]
Hey Amir,

Sorry for the late reply—things have been hectic with coursework and updates for the book.

I took a look at your atomic-autograd project, and I have to say, it’s a fantastic piece of work. I love how clean and intuitive it is. Projects like this often feel so simple (simple is a good thing!) and natural once you understand them, but teaching these ideas to others is where the real challenge lies. I’ve run into the same issue myself—finding the right balance between detailed explanations and keeping things straightforward isn’t easy, but your approach strikes a nice balance.

It’s great that you’re sharing this work, and I completely understand the difficulty of managing it alongside everything else. I’ve found that teaching helps me stay engaged, but I know how tough it can be to juggle everything. You'll be surprised how many hours and sleepless nights I pour into the https://mlsysbook.ai ( https://mlsysbook.ai/ ) and the tinytorch code. So I’m happy to support you however I can.

As for your idea about adding a CuPy GPU-backed implementation alongside NumPy, I think it’s a really compelling suggestion. The shared architecture and method signatures between the two libraries make it a super nice fit. I’ll spend some more time exploring your atomic-autograd implementation to see how we could integrate something similar. Multi-device execution is such an exciting area, especially as we think about scaling up to larger and more complex models. So I definitely want us to put that in. Thanks again for sharing your work.

Keep building and wishing you all the best!
